### PR TITLE
Import categories

### PR DESCRIPTION
**Pull request by SkySkimmer**
This PR allows to do eg `Import(coercions) X Y.` and `Import(-coercions) X Y.` meaning respectively "import only coercions" and "import everything except coercions".
Currently only the `coercions` category is defined. We may want to add a few more before merging.

Overlays:
- https://github.com/LPCIC/coq-elpi/pull/301

### CODE REVIEW

**Comment by Zimmi48** [comment]
>  Should the syntax be `Import (-cats) X` or `Import -(cats) X` for the negative version? (currently it's the first one)

It depends on how categories can be combined (and if there are sub and super categories).
With `-(cats)` this should mean eliminates all these categories.
With `-cats` the minus sign should be repeated if we want it to have the same meaning. If it's not repeated, then it could mean: add this category but removes this sub-category, but add this sub-sub-category, etc.

**Comment by Zimmi48** [comment]
If categories do not overlap and also do not overlap with the objects that can be imported with filtered imports, then I would favor the syntax with the minus sign outside the parenthese.

**Comment by SkySkimmer** [comment]
>If categories do not overlap and also do not overlap with the objects that can be imported with filtered imports, then I would favor the syntax with the minus sign outside the parentheses.

OK

**Comment by Alizter** [comment]
Also perhaps it would better to have an attribute for `Import` rather than modifying the syntax like this? Then we can have a small `categories` syntax in an attribute?
```
#[category(+coercions, -hints:core)]
Import Foo.
```

**Comment by Zimmi48** [comment]
As soon as we get the hints and the typeclass category though, there is some overlapping (e.g. I want to import all hints except typeclass intances).

File: doc/sphinx/language/core/modules.rst [hunk 1]
```diff
@@ -198,7 +198,7 @@ are now available through the dot notation.
    If :n:`@module_binder`\s are specified, declares a functor with parameters given by the list of
    :token:`module_binder`\s.
 
-.. cmd:: Import {+ @filtered_import }
+.. cmd:: Import {? @import_categories } {+ @filtered_import }
 
    .. insertprodn filtered_import filtered_import
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:203]:
```suggestion
   .. insertprodn import_categories filtered_import
```

This is to include the definition of `import_categories` in the manual.

Please add a sentence here describing what `import_categories` does.

**Review by jfehrle — DISMISSED** [review]
Minor tweak

**Comment by SkySkimmer** [comment]
@Alizter that's not current attribute syntax, what specific changes are you thinking about?

>As soon as we get the hints and the typeclass category though, there is some overlapping (e.g. I want to import all hints except typeclass intances).

I don't think we want to have overlapping categories. For TC we can treat it as a subcategory of hints.

-------

Some imaginary syntax
~~~coq
(* typeclasses is special, it means both the db and the other TC objects???? *)
Import (coercions, hints(core,typeclasses), notations(-("_ = _", "_ <= _")))
  X Y Z.
(* could also be "Import -(coercions, ...) X Y Z." *)

(* doesn't quite fit attribute syntax, strings are only allowed with att="string", but we could change that *)
#[only(coercions, hints(core,typeclasses), notations(except("_ = _", "_ <= _")))]
  Import X Y Z.

(* what if we want to import notation "_ = _" in scope bla but not the one for scope bli?? *)
Import (notations("_ = _"%bla)) (* would be very custom syntax, doesn't seem to fit attributes *)
#[only(notations(scoped("_ = _", bla))] (* ???? *)
~~~

**Comment by Alizter** [comment]
@SkySkimmer The second example you gave is what I had in mind

**Comment by ppedrot** [comment]
> in the form of a list of key=value pairs

That's literally what I call no structure. Thankfully we have invented type systems since LISP sexprs.

**Review by ejgallego — COMMENTED** [review]
Had a quick look, seems fine to me, I'll do a full review ASAP.

A quick note [not very important for this PR] is whether we want to have this effectful category API, or use instead an extensible category `ADT`; depends a bit what we'd like to do with libobject in the future, so maybe we want to think about that.

**Review by ejgallego — APPROVED** [review]

**Pull request merged**